students/JaeKim/lesson07/assignment/parallel.py

```diff
@@ -0,0 +1,186 @@
+import os
+import csv
+from pymongo import MongoClient
+import threading
+import datetime
+from multiprocessing.pool import ThreadPool
+from time import sleep
+
+
+class MongoDBConnection():
+    '''
+    Mongo connection
+    '''
+
+    def __init__(self, host='127.0.0.1', port=27017):
+        self.host = host
+        self.port = port
+        self.connection = None
+
+    def __enter__(self):
+        self.connection = MongoClient(self.host, self.port)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.connection.close()
+
+
+def cleanup():
+    '''
+    Used to cleanup mongo collections
+
+    :return:
+    '''
+    mongo = MongoDBConnection()
+
+    with mongo:
+        database = mongo.connection.media
+
+        [i.drop() for i in [database.products, database.customers, database.rentals]]
+
+
+def import_data(directory_name, product_file, rentals_file):
+    '''
+    Imports CSV files and uses threading to load them concurrently
+
+    :param directory_name:
+    :param product_file:
+    :param customer_file:
+    :param rentals_file:
+    :return:
+    '''
+    mongo = MongoDBConnection()
+
+    pool = ThreadPool(processes=3)
+
+    with mongo:
+        database = mongo.connection.media
+        products_csv, rentals_csv = (os.path.join(directory_name, filename) for filename in
+                                                   (product_file, rentals_file))
+
+        products_data = csv.reader(open(products_csv, encoding='utf-8-sig'))
+        rentals_data = csv.reader(open(rentals_csv, encoding='utf-8-sig'))
+
+        collection_products, collection_rentals = database['products'], database['rentals']
+        data = [products_data, rentals_data]
+
+        records_inserted = [pool.apply_async(_insert_data, (collection, data)).get() for collection, data in
+                   zip([collection_products, collection_rentals], data)]
+
+        print(records_inserted)
+
+        pool.close()
+        pool.join()
+
+    return records_inserted
```

Comment:
Did it work? Did you speed it up any? It is dependent on how mongodb is implemented. If it only takes one connection at a time, then it won't matter if we thread. If these were network calls, we would almost certainly see faster execution times, but since mongodb is local, not guaranteed.